MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m

```diff
@@ -0,0 +1,468 @@
+                [self handleCancelEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationAccept:
+                [self handleAcceptEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationKey:
+                [self handleKeyEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationMac:
+                [self handleMacEvent:event];
+                break;
+
+            default:
+                break;
+        }
+    });
+}
+
+- (void)handleStartEvent:(MXEvent*)event
+{
+    NSLog(@"[MXKeyVerification] handleStartEvent");
+
+    MXKeyVerificationStart *keyVerificationStart;
+    MXJSONModelSetMXJSONModel(keyVerificationStart, MXKeyVerificationStart, event.content);
+
+    if (!keyVerificationStart)
+    {
+        return;
+    }
+    if (!keyVerificationStart.isValid)
+    {
+        if (keyVerificationStart.transactionId && keyVerificationStart.fromDevice)
+        {
+            [self cancelTransaction:keyVerificationStart.transactionId
+                         fromUserId:event.sender
+                          andDevice:keyVerificationStart.fromDevice
+                               code:MXTransactionCancelCode.invalidMessage];
+        }
+
+        return;
+    }
+
+
+    MXDeviceVerificationTransaction *existingTransaction = [self transactionWithUser:keyVerificationStart.transactionId andDevice:keyVerificationStart.fromDevice];
+    if (existingTransaction)
+    {
+        NSLog(@"[MXKeyVerification] handleStartEvent: already existing transaction. Cancel both");
+
+        [existingTransaction cancelWithCancelCode:MXTransactionCancelCode.invalidMessage];
+        [self cancelTransaction:keyVerificationStart.transactionId
+                     fromUserId:event.sender
+                      andDevice:keyVerificationStart.fromDevice
+                           code:MXTransactionCancelCode.invalidMessage];
+        return;
+    }
+
+
+    // Make sure we have other device keys
+    [self loadDeviceWithDeviceId:keyVerificationStart.fromDevice andUserId:event.sender success:^(MXDeviceInfo *otherDevice) {
+
+        // Multiple keyshares between two devices: any two devices may only have at most one key verification in flight at a time.
+        NSArray<MXDeviceVerificationTransaction*> *transactionsWithUser = [self transactionsWithUser:event.sender];
+        if (transactionsWithUser)
+        {
+            NSLog(@"[MXKeyVerification] handleStartEvent: already existing transaction with the user. Cancel both");
+
+            [existingTransaction cancelWithCancelCode:MXTransactionCancelCode.invalidMessage];
+            [self cancelTransaction:keyVerificationStart.transactionId
+                         fromUserId:event.sender
+                          andDevice:keyVerificationStart.fromDevice
+                               code:MXTransactionCancelCode.invalidMessage];
+            return;
+        }
+
+        
+        // We support only SAS at the moment
```

Comment:
Should test that method == "m.sas.v1" ?

Reply:
>  Oh it's done by the initWithOtherDevice?

yes. This is part of the validation process within `MXIncomingSASTransaction.initWithOtherDevice()`